features/field-based-acl.md

```diff
@@ -0,0 +1,243 @@
+resolve order will be:
+
+1. **Owner**: the owner of the record instance
+1. **Specific User**: the user specified by user ID
+1. **Reference Users**: the users specified in a field of the records
+1. **Defined Roles**: the user-defined roles
+1. **Any Users**: any logged-in users
+1. **Public**: any users with correct API key
+
+## Base ACL
+
+Since the field-based ACL is allow-based, the following entities would be added
+serving as a base ACL.
+
+| Class | UserRole | Field | AccessLevel | DiscoveryLevel |
+|-------|----------|-------|-------------|----------------|
+| *     | Public   | *     | ReadWrite   | Queryable      |
+
+## Work together with record-based ACL
+
+The field-based ACL would be working together with the record-based ACL.
+Record-based ACL controls whether the user can access a record. Field-based ACL
+controls whether the user can access a field of a record. When a user is
+operating on some fields of some records, it would be tested by both
+record-based ACL and field-based ACL.
+
+The following pseudo code shows how an record operation being tested by both
+record-based ACL and field-based ACL:
+
+```js
+function recordOperationVeridation (operation) {
+  const currentUser = global.context.getUser();
+  const originalRecord = operation.originalRecord;
+  const operationType = operation.operationType;
+
+  const recordAccess = originalRecord.access.hasAccess(
+    operationType,
+    currentUser
+  );
+  if (recordAccess == false) {
+    throw new NoAccessError(currentUser, operationType);
+  }
+
+  const recordFieldAccess = originalRecord.class.fieldAccess;
+  const operatingFields = operation.operatingFields;
+
+  // not considered access depends on data of the record
+  operatingFields.forEach(function (eachOperatingField) {
+    const eachOperatingFieldAccess = recordFieldAccess.hasAccess(
+      eachOperatingField,
+      operationType,
+      currentUser
+    );
+    if (eachOperatingFieldAccess == false) {
+      throw new NoAccessError(currentUser, operationType, eachOperatingField);
+    }
+  });
+
+  return true;
+}
+```
+
+# Sample ACL for Use Cases
+
+- Make gender field of user record private to every one except owner
+
+| Class | UserRole | Field  | AccessLevel | DiscoveryLevel |
+|-------|----------|--------|-------------|----------------|
+| *     | Public   | *      | ReadWrite   | Queryable      |
+| User  | AnyUser  | gender | NoAccess    | NotQueryable   |
+| User  | Owner    | gender | ReadWrite   | Queryable      |
+
+- Make gender field of user record private to every one, readable and queryable
+  to friends and updatable to owner
+
+| Class |  UserRole   | Field  | AccessLevel | DiscoveryLevel |
+|-------|-------------|--------|-------------|----------------|
+| *     | Public      | *      | ReadWrite   | Queryable      |
+| User  | AnyUser     | gender | NoAccess    | NotQueryable   |
+| User  | Ref:Friends | gender | ReadOnly    | Queryable      |
```

Comment:
`Ref:Friends` need update to clarify it is not related to friends relation

perhaps: `col:group_members`?